src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -0,0 +1,203 @@
+    | CallLikeExpression ({Expression = Identifier (GlobalCallable qualName, _)}, b)
+        when (cd.getCallable qualName).Kind = TypeConstructor -> isLiteral b.Expression cd
+    | CallLikeExpression (a, b) ->
+        isLiteral a.Expression cd && isLiteral b.Expression cd &&
+            TypedExpression.IsPartialApplication (CallLikeExpression (a, b))
+    | _ -> false
+
+
+/// Helper method to improve expression readability
+let rec prettyPrint (expr: Expr): string =
+    match expr with
+    | Identifier (LocalVariable a, _) -> "LocalVariable " + a.Value
+    | Identifier (GlobalCallable a, _) -> "GlobalCallable " + a.Name.Value
+    | CallLikeExpression (f, x) -> "(" + (prettyPrint f.Expression) + " of " + (prettyPrint x.Expression) + ")"
+    | ValueTuple a -> "(" + String.Join(", ", a |> Seq.map (fun x -> x.Expression) |> Seq.map prettyPrint) + ")"
+    | ValueArray a -> "[" + String.Join(", ", a |> Seq.map (fun x -> x.Expression) |> Seq.map prettyPrint) + "]"
+    | ADD (a, b) -> "(" + (prettyPrint a.Expression) + " + " + (prettyPrint b.Expression) + ")"
+    | a -> a.ToString()
+
+
+/// Converts a range literal to a sequence of integers
+let rangeLiteralToSeq (r: Expr): seq<int> =
+    match r with
+    | RangeLiteral (a, b) ->
+        match a.Expression, b.Expression with
+        | IntLiteral start, IntLiteral stop ->
+            seq { int start .. int stop }
+        | RangeLiteral ({Expression = IntLiteral start}, {Expression = IntLiteral step}), IntLiteral stop ->
+            seq { int start .. int step .. int stop }
+        | _ -> failwithf "Invalid range literal: %O" r
+    | _ -> failwithf "Not a range literal: %O" r
+
+
+/// Represents the current state of all the local variables in a function
+type VariablesDict() =
+    let scopeStack = new Stack<Dictionary<string, Expr>>()
+
+    member this.enterScope(): unit =
+        scopeStack.Push (new Dictionary<string, Expr>()) |> ignore
+
+    member this.exitScope(): unit =
+        // TODO: assert scopeStack is nonempty
+        scopeStack.Pop () |> ignore
+
+    member this.defineVar(name: string, value: Expr): unit =
+        // TODO: assert variable is undefined
+        scopeStack.Peek().[name] <- value
+
+    member this.setVar(name: string, value: Expr): unit =
+        // TODO: assert variable is defined, is same type
+        (scopeStack |> Seq.find (fun x -> x.ContainsKey name)).[name] <- value
+
+    member this.getVar(name: string): Expr option =
+        match scopeStack |> Seq.tryFind (fun x -> x.ContainsKey name) with
+        | Some dict -> Some dict.[name]
+        | None -> None
+
+    override this.ToString(): string =
+        "[" + String.Join("\n", scopeStack |> Seq.map (fun x ->
+            "{" + String.Join(", ", x |> Seq.map (fun y -> y.Key + "=" + y.Value.ToString())) + "}"
+        )) + "]"
+
+        
+/// Represents a (possibly-nested) tuple of strings.
+/// Used as a way to compare QsTuples and SymbolTuples.
+type StringTuple =
```

Comment:
You could instead use the ITuple interface, which serves pretty much that purpose. 
Otherwise please make this type private or internal, since making things public means changes to it are breaking changes (i.e. you are implicitly giving at least some stability and maintenance guarantees).
The same goes for a lot of the used data structures here - things that are publicly accessible are things that cannot be changed easily.